## v4.0
## Краткий список изменений
- Изменена работа мидлварей
- Изменен аргумент в рулзах
- Новый синтаксис для `CodeException`
- Поддержка юзерботов

## Изменения в мидлварях
Теперь методы `pre` и `post` не принимают аргумент message.\
Доступ к нему теперь происходит через `self.event`, а его тип указывается в дженерике.

[Подробнее](../high-level/handling/middleware.md)

## Изменения в рулзах
Аргумент message переименован в event

## Новый синтаксис для `CodeException`
Теперь ошибка в `CodeException` указывается в квадратных скобках.\
Это касается и `VKAPiError`:
```python
async def main():
    try:
        await api.users.get(user_ids=["123456789"])
    except VKAPIError[5]:
        print("Ой, неверный ключ доступа.")
    except VKAPIError[30]:
        print("Ой, у пользователя закрытый профиль.")
    except VKAPIError as e:
        print(f"Произошла ошибка {e.code}.")
```
[Подробнее](../low-level/exception_handling/code-exception.md)

## Поддержка юзерботов
Добавлен класс `User`, все аналогично обычным ботам.\
Импортируется из `vkbottle.user`
